### PR TITLE
Core: Support valid globs

### PR DESCRIPTION
**Pull request by ndelangen**
Issue: https://github.com/storybookjs/storybook/issues/10772

## What I did
The value of `stories` in `main.js` is supposed to be a valid glob. We plan to optimize storybook by moving from a regex evaluated in webpack to a glob evaluated by node.

But we made a mistake, when we wrote the original globs in our docs & CLI generators. It's actually an invalid glob, but it still works, because the glob-to-regex lib makes it work.

In the future instead of using a glob-to-regex we'll use glob solely.

I added instructions to migration.md & make the warning clearer on what's really happening.

We'll keep the invalid globs working in storybook for now, but they will break at some point in the future.

### CODE REVIEW

**Review by shilman — COMMENTED** [review]
Does "at some point in the future" mean 7.0 or later?

If you want to do that optimization in a 6.x release, we should make the breaking change now. WDYT?

File: MIGRATION.md [hunk 1]
```diff
@@ -158,6 +159,24 @@ Storybook has built-in Typescript support in 6.0. That means you should remove y
...
+We've corrected this, the CLI templates have been changed to valid globs.
+
+We've also changed the code that resolves these globs, so that invalid globs will log a warning. They will break in the future, so if you see this warning, please ensure you're specifying a valid glob.
+
+Example of an **invalid** glob:
+```
+stories: ['./**/*.stories.(ts|js)']
+```
+
+Example of an **valid** glob:
```
Comment (yannbf) [inline MIGRATION.md:175]:
```suggestion
Example of a **valid** glob:
```

File: MIGRATION.md [hunk 2]
```diff
@@ -158,6 +159,24 @@ Storybook has built-in Typescript support in 6.0. That means you should remove y
 
 To migrate from an old setup, we recommend deleting any typescript-specific webpack/babel configurations in your project. If you want to override the defaults, see the [typescript configuration docs](https://github.com/storybookjs/storybook/blob/next/docs/src/pages/configurations/typescript-config/index.md).
 
+### Correct globs in main.js
+
+In 5.3 we introduced the `main.js` file with an `stories` property. This property was documented as a "glob" pattern. This was our intention, however the implementation allowed for non valid globs to be specified and work. In fact we promoted invalid globs in our documentation and CLI templates.
```
Comment (yannbf) [inline MIGRATION.md:164]:
```suggestion
In 5.3 we introduced the `main.js` file with a `stories` property. This property was documented as a "glob" pattern. This was our intention, however the implementation allowed for non valid globs to be specified and work. In fact, we promoted invalid globs in our documentation and CLI templates.
```

File: MIGRATION.md [hunk 3]
```diff
@@ -158,6 +159,24 @@ Storybook has built-in Typescript support in 6.0. That means you should remove y
...
+
+We've corrected this, the CLI templates have been changed to valid globs.
```
Comment (yannbf) [inline MIGRATION.md:166]:
```suggestion
We've corrected this, the CLI templates have been changed to use valid globs.
```

File: lib/core/src/server/preview/to-require-context.test.js [hunk 4]
```diff
@@ -46,6 +47,79 @@ const testCases = [
...
+    invalidPaths: [
+      './stories.js',
+      './src/stories/Icon.stories.js',
+      './Icon.stories.js',
+      '../src/Icon.stories.mdx',
+      '../src/stories/components/Icon/Icon.stories.ts',
+      '../src/stories/components/Icon/Icon.mdx',
+    ],
+  },
+  // DUMB GLOB
```
Comment (yannbf) [inline lib/core/src/server/preview/to-require-context.test.js:104]: should this comment stay here? 😂

**Review by yannbf — APPROVED** [review]
I'm so happy this is fixed <3

**Comment by ndelangen** [comment]
So turns out, valid globs break webpack. I've investigating why, but it's not going well.

File: lib/core/src/server/preview/to-require-context.js [hunk 5]
```diff
@@ -22,5 +33,8 @@ export const toRequireContext = (input) => {
 
 export const toRequireContextString = (input) => {
   const { path: p, recursive: r, match: m } = toRequireContext(input);
-  return `require.context('${p}', ${r}, ${m})`;
+
+  const result = `console.log('HERE') || require.context('${p}', ${r}, /${m}/)`;
+  // process.exit(0);
+  return result;
```
Comment (shilman) [inline lib/core/src/server/preview/to-require-context.js:39]: FIXME

**Review by shilman — CHANGES_REQUESTED** [review]
Looks good except for the FIXME.

Also can you provide a handful unit tests for this since it's complex and deserves a little documentation

**Review by shilman — APPROVED** [review]
Nice! Didn't realize there were already unit tests in the other file 👍

File: lib/core/src/server/preview/to-require-context.js [hunk 6]
```diff
@@ -1,14 +1,37 @@
 import globBase from 'glob-base';
-import { makeRe } from 'nanomatch';
+import { makeRe } from 'micromatch';
+
+// LEGACY support for bad glob patterns we had in SB 5 - remove in SB7
+const fixBadGlob = (val) => {
+  const match = val.match(/\.(\([^)]+\))/);
+
+  if (match) {
+    return match.input.replace(match[1], `@${match[1]}`);
```
Comment (shilman) [inline lib/core/src/server/preview/to-require-context.js:9]: Can we add a `util-deprecate` deprecation warning here?

File: lib/core/src/server/preview/to-require-context.js [hunk 7]
```diff
@@ -22,5 +45,8 @@ export const toRequireContext = (input) => {
 
 export const toRequireContextString = (input) => {
   const { path: p, recursive: r, match: m } = toRequireContext(input);
-  return `require.context('${p}', ${r}, ${m})`;
+
+  const result = `require.context('${p}', ${r}, /${m}/)`;
+  // process.exit(0);
```
Comment (shilman) [inline lib/core/src/server/preview/to-require-context.js:50]:
```suggestion
```

**Review by shilman — CHANGES_REQUESTED** [review]
Couple more things

File: lib/core/src/server/preview/to-require-context.js (hunk 6 above)
Comment (ndelangen) [reply lib/core/src/server/preview/to-require-context.js:9]: good call, will add it

**Comment by ndelangen** [comment]
So happy to have a found a solution, and a clean break on how to deal with the invalid globs in a backwards but future friendly way.

All we need to do to remove the bad behavior support is delete a few lines of code.

File: addons/docs/docs/docspage.md [hunk 8]
```diff
@@ -155,7 +155,7 @@ You can interleave your own components to customize the auto-generated contents
 
 ## Story file names
 
-Unless you use a custom webpack configuration, all of your story files should have the suffix `*.stories.[jt]sx?`, e.g. `"Badge.stories.js"`, `"Badge.stories.tsx"`, etc.
+Unless you use a custom webpack configuration, all of your story files should have the suffix `*.stories.@(j|t)]sx?`, e.g. `"Badge.stories.js"`, `"Badge.stories.tsx"`, etc.
```
Comment (shilman) [inline addons/docs/docs/docspage.md:158]:
```suggestion
Unless you use a custom webpack configuration, all of your story files should have the suffix `*.stories.@(j|t)sx?`, e.g. `"Badge.stories.js"`, `"Badge.stories.tsx"`, etc.
```

File: lib/core/src/server/preview/to-require-context.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+import deprecate from 'util-deprecate';
+import dedent from 'ts-dedent';
+
+// LEGACY support for bad glob patterns we had in SB 5 - remove in SB7
+const fixBadGlob = deprecate(
+  (match: RegExpMatchArray) => {
+    return match.input.replace(match[1], `@${match[1]}`);
+  },
+  dedent`
+    You're specified an invalid glob, we've attempted to fix it, please ensure the glob you specify is a valid glob
```
Comment (shilman) [inline lib/core/src/server/preview/to-require-context.ts:12]:
```suggestion
    You're specified an invalid glob, we've attempted to fix it, please ensure the glob you specify is a valid glob. See: https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#correct-globs-in-main-js
```

**Review by shilman — APPROVED** [review]
Two more 😅

File: lib/core/src/server/preview/entries.js [hunk 10]
```diff
@@ -57,13 +57,18 @@ export async function createPreviewEntry(options) {
   if (stories && stories.length) {
     entries.push(path.resolve(path.join(configDir, `generated-stories-entry.js`)));
 
-    const files = (await Promise.all(stories.map((g) => glob(g)))).reduce((a, b) => a.concat(b));
+    const files = (
+      await Promise.all(stories.map((g) => glob(path.join(configDir, g))))
```
Comment (hipstersmoothie) [inline lib/core/src/server/preview/entries.js:61]: @ndelangen Is this correct? I don't think it makes sense to join the glob with the configDir. This means that all globs before this pr won't work anymore

Comment (ndelangen) [reply lib/core/src/server/preview/entries.js:61]: yes this is intentional indeed.

Comment (shilman) [reply lib/core/src/server/preview/entries.js:61]: @ndelangen if this is a breaking change do we need to document it in `MIGRATION.md`?

Comment (hipstersmoothie) [reply lib/core/src/server/preview/entries.js:61]: This is def a breaking change that should be documented. But it seems really odd to me that your glob become `.storybook/**/*.stories`. IDK when I would ever want that. In reality it always must be something like `.storybook/../**/*.stories`.

**Pull request merged**